internal/clients/kafka/topic/topic.go

```diff
@@ -0,0 +1,192 @@
+package topic
+
+import (
+	"context"
+	//"github.com/crossplane/crossplane-runtime/pkg/meta"
+	//"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	//"github.com/twmb/franz-go/pkg/kerr"
+
+	"github.com/pkg/errors"
+	"github.com/twmb/franz-go/pkg/kadm"
+
+	"github.com/crossplane-contrib/provider-kafka/apis/topic/v1alpha1"
+)
+
+// Topic is a holistic representation of a Kafka Topic with all configurable
+// fields
+type Topic struct {
+	Name              string
+	ReplicationFactor int16
+	Partitions        int32
+	ID                string
+	Config            map[string]*string
+}
+
+// Get gets the topic from Kafka side and returns a Topic object.
+func Get(ctx context.Context, client *kadm.Client, name string) (*Topic, error) {
+
+	td, err := client.ListTopics(ctx, name)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot list topics")
+	}
+	if td[name].Err != nil {
+		return nil, errors.Wrap(td[name].Err, "topic does not exist in kafka cluster")
+	}
+
+	t, ok := td[name]
+	if !ok {
+		return nil, errors.New("no create response for topic")
+	}
+
+	tc, err := client.DescribeTopicConfigs(ctx, name)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot describe topics")
+	}
+
+	ts := Topic{}
+	ts.Name = name
+	ts.ReplicationFactor = int16(t.Partitions[0].Replicas[0])
+	ts.Partitions = t.Partitions[0].Partition
+	ts.ID = t.ID.String()
+	ts.Config = make(map[string]*string)
+
+	for _, value := range tc[0].Configs {
+		ts.Config[value.Key] = value.Value
+	}
+	return &ts, nil
+
+}
+
+func Create(ctx context.Context, client *kadm.Client, topic *Topic) error {
+
+	resp, err := client.CreateTopics(ctx, topic.Partitions, topic.ReplicationFactor, topic.Config, topic.Name)
+	if err != nil {
+		return err
+	}
+
+	t, ok := resp[topic.Name]
+	if !ok {
+		return errors.New("no create response for topic")
+	}
+	if t.Err != nil {
+		return errors.Wrap(t.Err, "cannot create topic")
+	}
+
+	return nil
+}
+
+// Delete deletes the topic from Kafka side
+func Delete(ctx context.Context, client *kadm.Client, name string) error {
+
+	td, err := client.DeleteTopics(ctx, name)
+	if err != nil {
+		return err
+	}
+
+	t, ok := td[name]
+	if !ok {
+		return errors.New("no delete response for topic")
+	}
+	if t.Err != nil {
+		return errors.Wrap(t.Err, "cannot delete topic")
+	}
+
+	return nil
+}
+
+func Update(ctx context.Context, client *kadm.Client, desired *Topic) error {
+	// First Get existing Topic
+	existing, err := Get(ctx, client, desired.Name)
+	if err != nil {
+		return errors.Wrap(err, "cannot get topic")
+	}
+	if existing == nil {
+		return errors.New("topic does not exist")
+	}
+
+	_, err = client.UpdatePartitions(ctx, int(desired.Partitions), desired.Name)
+	if err != nil {
+		return errors.Wrap(err, "cannot update topic partitions")
+	}
+
+	if desired.Config != nil {
+		configs := desired.Config
+		for key, value := range configs {
```

Comment:
Since we already have `existing.Config`, we should better only update if different.